### PR TITLE
Rebase IWA work and small fixes

### PR DESCRIPTION
**Pull request by bgavrilMS**
Please review only last 2 PRs, the rest have been reviewed.

### CODE REVIEW

File: core/src/Realm/UserRealmDiscoveryResponse.cs [hunk 1]
```diff
@@ -53,18 +54,18 @@ internal sealed class UserRealmDiscoveryResponse
...
 
+        internal static async Task<UserRealmDiscoveryResponse> CreateByDiscoveryAsync(string userRealmUriPrefix, string userName, RequestContext requestContext)
+        {
             var msg = "Sending request to userrealm endpoint.";
             requestContext.Logger.Info(msg);
             requestContext.Logger.InfoPii(msg);
 
-            var client = new AdalHttpClient(userRealmEndpoint, requestContext) { Client = { Accept = "application/json" } };
-            return await client.GetResponseAsync<UserRealmDiscoveryResponse>().ConfigureAwait(false);
+            Debug.Assert(userRealmUriPrefix.EndsWith("/"), "Expecting the userRealmUriPrefix to end in / so that it can be appended to");
```
Comment (henrik-me) [inline core/src/Realm/UserRealmDiscoveryResponse.cs]: >Debug.Assert [](start = 12, length = 12)

remove, if needed for debugging add logging.

File: msal/tests/dev apps/DesktopTestApp/MainForm.cs [hunk 2]
```diff
@@ -81,11 +82,11 @@ public void LogDelegate(LogLevel level, string message, bool containsPii)
         public void RefreshUserList()
         {
             List<IAccount> accounts = _publicClientHandler.PublicClientApplication.GetAccountsAsync().Result.ToList();
-            if (accounts.Count > 0)
-            {
-                accounts.Insert(0, new Account(_publicClientHandler.CurrentUser.HomeAccountId, 
-                    _publicClientHandler.CurrentUser.Username, _publicClientHandler.CurrentUser.Environment));
-            }
+            //if (accounts.Count > 0)
```
Comment (henrik-me) [inline msal/tests/dev apps/DesktopTestApp/MainForm.cs]: >//if (accounts.Count > 0) [](start = 12, length = 25)

why is this removed?

**Review by henrik-me — APPROVED** [review]
:shipit:

File: msal/src/Microsoft.Identity.Client/PublicClientApplication.cs [hunk 3]
```diff
@@ -418,6 +418,49 @@ public async Task<AuthenticationResult> AcquireTokenAsync(IEnumerable<string> sc
...
+        /// <param name="scopes">Array of scopes requested for resource</param>
+        /// <returns>Authentication result containing token of the current login user</returns>
+        public async Task<AuthenticationResult> AcquireTokenByIntegratedWindowsAuthAsync(IEnumerable<string> scopes)
+        {
+            //TODO: make it platform specific
+            return await AcquireTokenByIWAAsync(scopes, new IWAInput()).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Non-interactive request to acquire token via Windows Integrated Authentication.
```
Comment (jmprieur) [inline msal/src/Microsoft.Identity.Client/PublicClientApplication.cs]: <summary>Non-interactive request to acquire a security token for the signed-in user in Windows, via Integrated Windows Authentication. See https://aka.ms/msal-net-iwa.</summary>
<remark>This overrides enables the application developer to provide a hint about the logged-in user, as on some windows installation the windows administrator might have prevented the applications from accessing the identity of the logged-in user</remark>

File: msal/src/Microsoft.Identity.Client/PublicClientApplication.cs [hunk 4]
```diff
@@ -418,6 +418,49 @@ public async Task<AuthenticationResult> AcquireTokenAsync(IEnumerable<string> sc
                         behavior, extraQueryParameters, parent, ApiEvent.ApiIds.AcquireTokenWithScopeUserBehaviorAuthority).ConfigureAwait(false);
         }
 
+        /// <summary>
+        /// Non-interactive request to acquire token via Kerberos / Windows Integrated Authentication.
+        /// The username for which the flow is executed is pulled from the operating system, as the current user principal name
+        /// </summary>
+        /// <param name="scopes">Array of scopes requested for resource</param>
```
Comment (jmprieur) [inline msal/src/Microsoft.Identity.Client/PublicClientApplication.cs]: Scopes requested to access a protected API

File: msal/src/Microsoft.Identity.Client/PublicClientApplication.cs [hunk 5]
```diff
@@ -418,6 +418,49 @@ public async Task<AuthenticationResult> AcquireTokenAsync(IEnumerable<string> sc
...
+        public async Task<AuthenticationResult> AcquireTokenByIntegratedWindowsAuthAsync(IEnumerable<string> scopes)
+        {
+            //TODO: make it platform specific
+            return await AcquireTokenByIWAAsync(scopes, new IWAInput()).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Non-interactive request to acquire token via Windows Integrated Authentication.
+        /// </summary>
+        /// <param name="scopes">Array of scopes requested for resource</param>
```
Comment (jmprieur) [inline msal/src/Microsoft.Identity.Client/PublicClientApplication.cs]: Scopes requested to access a protected API

File: msal/src/Microsoft.Identity.Client/PublicClientApplication.cs [hunk 6]
```diff
@@ -418,6 +418,49 @@ public async Task<AuthenticationResult> AcquireTokenAsync(IEnumerable<string> sc
                         behavior, extraQueryParameters, parent, ApiEvent.ApiIds.AcquireTokenWithScopeUserBehaviorAuthority).ConfigureAwait(false);
         }
 
+        /// <summary>
+        /// Non-interactive request to acquire token via Kerberos / Windows Integrated Authentication.
+        /// The username for which the flow is executed is pulled from the operating system, as the current user principal name
+        /// </summary>
+        /// <param name="scopes">Array of scopes requested for resource</param>
+        /// <returns>Authentication result containing token of the current login user</returns>
```
Comment (jmprieur) [inline msal/src/Microsoft.Identity.Client/PublicClientApplication.cs]: Authentication result containing a token for the requested scopes and for the currently logged-in user in Windows

File: msal/src/Microsoft.Identity.Client/PublicClientApplication.cs [hunk 7]
```diff
@@ -418,6 +418,49 @@ public async Task<AuthenticationResult> AcquireTokenAsync(IEnumerable<string> sc
...
+        {
+            //TODO: make it platform specific
+            return await AcquireTokenByIWAAsync(scopes, new IWAInput()).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Non-interactive request to acquire token via Windows Integrated Authentication.
+        /// </summary>
+        /// <param name="scopes">Array of scopes requested for resource</param>
+        /// <param name="username">Username for which to run the flow</param>
```
Comment (jmprieur) [inline msal/src/Microsoft.Identity.Client/PublicClientApplication.cs]: Identifier of the user account for which to acquire a token with Integrated Windows authentication. Generally in UserPrincipalName (UPN) format, e.g. john.doe@contoso.com

File: msal/src/Microsoft.Identity.Client/PublicClientApplication.cs [hunk 8]
```diff
@@ -418,6 +418,49 @@ public async Task<AuthenticationResult> AcquireTokenAsync(IEnumerable<string> sc
...
+        /// <returns>Authentication result containing token of the specified login user</returns>
```
Comment (jmprieur) [inline msal/src/Microsoft.Identity.Client/PublicClientApplication.cs]: Authentication result containing a token for the requested scopes and for the currently logged-in user in Windows

File: msal/src/Microsoft.Identity.Client/PublicClientApplication.cs [hunk 9]
```diff
@@ -418,6 +418,49 @@ public async Task<AuthenticationResult> AcquireTokenAsync(IEnumerable<string> sc
                         behavior, extraQueryParameters, parent, ApiEvent.ApiIds.AcquireTokenWithScopeUserBehaviorAuthority).ConfigureAwait(false);
         }
 
+        /// <summary>
+        /// Non-interactive request to acquire token via Kerberos / Windows Integrated Authentication.
```
Comment (jmprieur) [inline msal/src/Microsoft.Identity.Client/PublicClientApplication.cs]: Non-interactive request to acquire a security token for the signed-in user in Windows, via Integrated Windows Authentication. See https://aka.ms/msal-net-iwa.
The account used in this overrides is pulled from the operating system as the current user principal name

**Review by jmprieur — CHANGES_REQUESTED** [review]
LGTM. works (I've tested it)

I'd like you to change the XML comments as indicated.
Thanks!

File: msal/tests/dev apps/DesktopTestApp/MainForm.cs (hunk 2 above)
Comment (bgavrilMS) [reply msal/tests/dev apps/DesktopTestApp/MainForm.cs]: Not sure what it was doing before moving to accounts, but now it just duplicates an account. The UI is already bound to accounts variable.
